### PR TITLE
Workaround to support ruby as an interpreter

### LINKED ISSUE #877: nix-shell shebang and Ruby
My version of Nix is 1.11.2

The following script

```ruby
#! /usr/bin/env nix-shell
#! nix-shell -i ruby -p ruby

puts "Hey Nix!"
```

 ends with an error:

```
% ./foo
/nix/store/i9mr167d0b3s7n55q4yydnkvnnzq2hv2-ruby-2.3.0-p0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- uby (LoadError)
	from /nix/store/i9mr167d0b3s7n55q4yydnkvnnzq2hv2-ruby-2.3.0-p0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
zsh: exit 1     ./foo
```

This seems to be caused by nix-shell trying to execute ruby like this: `ruby -ruby ./foo` for some reason, as that gives the exact same error:

```
% ruby -ruby foo                                                                                                                                                                                                          17:05:42
/nix/store/i9mr167d0b3s7n55q4yydnkvnnzq2hv2-ruby-2.3.0-p0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- uby (LoadError)
	from /nix/store/i9mr167d0b3s7n55q4yydnkvnnzq2hv2-ruby-2.3.0-p0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
zsh: exit 1     ruby -ruby foo
```

I've looked a bit through the source and found that nix-shell also uses an hack to work around Perl shebang peculiarities, and my first thought was that maybe Ruby copied them from Perl.

Unfortunately I couldn't test my theory by patching nix because compilation fails for me (which I'll file in another issue).

### PR DESCRIPTION
**Pull request by sheenobu**
Fixes #877

### CODE REVIEW

**Comment by sheenobu** [comment]
require and require_relative were attempted but they will only load if the file ends in '.rb'.

**Comment by manveru** [comment]
@edolstra you think this is good to merge or is there a better way?

File: scripts/nix-build.in [hunk 1]
```diff
@@ -191,17 +191,34 @@ for (my $n = 0; $n < scalar @ARGV; $n++) {
...
             my $s = $_;
             $s =~ s/'/'\\''/g;
             return "'" . $s . "'";
         }
-        $envCommand = "exec $execArgs $interpreter $script ${\(join ' ', (map shellEscape, @savedArgs))}";
+
+        if ($interpreter =~ /ruby/) {
+
+            # Hack for Ruby. Ruby also examines the shebang. It tries to
+            # read the shabang to understand which packages to read from. Since
```
Comment (zimbatm) [inline scripts/nix-build.in]: typo: shabang -> shebang

**Comment by edolstra** [comment]
Seems good to me though I don't know anything about Ruby.

File: scripts/nix-build.in [hunk 2]
```diff
@@ -191,17 +191,34 @@ for (my $n = 0; $n < scalar @ARGV; $n++) {
...
-        $envCommand = "exec $execArgs $interpreter $script ${\(join ' ', (map shellEscape, @savedArgs))}";
+
+        if ($interpreter =~ /ruby/) {
+
+            # Hack for Ruby. Ruby also examines the shebang. It tries to
+            # read the shebang to understand which packages to read from. Since
+            # this is handled via nix-shell -p, we wrap our ruby script execution
+            # in an eval.
+            my $cmd = "$interpreter -e 'eval(File.read(\"$script\"), nil, \"$script\")'";
+            $envCommand = "exec $execArgs $cmd ${\(join ' ', (map shellEscape, @savedArgs))}";
```
Comment (zimbatm) [inline scripts/nix-build.in]:
```perl
$envCommand = "exec $execArgs $interpreter -e 'load(\"$script\")' -- ${\(join ' ', (map shellEscape, @savedArgs))}";
```

`--` is important if you don't want the arguments to be passet to the ruby interpreter

File: scripts/nix-build.in [hunk 3]
```diff
@@ -191,17 +191,34 @@ for (my $n = 0; $n < scalar @ARGV; $n++) {
...
+
         # Überhack to support Perl. Perl examines the shebang and
         # executes it unless it contains the string "perl" or "indir",
         # or (undocumented) argv[0] does not contain "perl". Exploit
         # the latter by doing "exec -a".
-        my $execArgs = $interpreter =~ /perl/ ? "-a PERL" : "";
+
+        if ($interpreter =~ /perl/) {
+            $execArgs = "-a PERL";
+        }
```
Comment (zimbatm) [inline scripts/nix-build.in]: move this closer to the other $interpreter check ?

File: scripts/nix-build.in (hunk 2 above)
Comment (sheenobu) [inline scripts/nix-build.in]: I'll use your example but the third argument to ruby eval should set $0 correctly. Maybe. You know i'm not exactly sure. Load is certainly cleaner, thank you.

Comment (sheenobu) [inline scripts/nix-build.in]: @zimbatm taken care of.

File: scripts/nix-build.in (hunk 3 above)
Comment (sheenobu) [inline scripts/nix-build.in]: Taken care of

**Comment by zimbatm** [comment]
All good now

**Comment by manveru** [comment]
@edolstra any chance you could merge this before the next release? :)

**Comment by Mic92** [comment]
Could this bugfix be cherry-picked?

**Comment by xzfc** [comment]
Probably I am late here, but I think the proper way of fixing this is to put `#! ruby` into the script before `#! nix-shell` line.

```ruby
#! /usr/bin/env nix-shell
#! ruby
#! nix-shell -i ruby -p ruby

puts "Hey Nix!"
```

It doesn't require such a hack in `nix-shell` code, and more importantly, it is compatible with vanilla ruby (that is, you can run `ruby ./foo.rb` without using `nix-shell`).

**Pull request merged**